### DISCUSSION TITLE
Question on replacing a color to transparent

### DISCUSSION THREAD

**Discussion by alagiz** [libvips/libvips#2474] [Q&A] [answered]
Hi!

I am trying to replace a color (black background) to transparent, but didn't succeed yet.
would appreciate any help/suggestions!
here is my code (it is based on this answer https://stackoverflow.com/a/55629222):

```python
import pyvips

image = pyvips.Image.new_from_file("baseImage.png", access='sequential')

# color to make transparent
colorToReplace = [0, 0, 0, 255]

# tolerance
tolerance = 25

# for each pixel, pythagorean distance from target colour
distance = sum(((image - colorToReplace) ** 2).bandsplit()) ** 0.5

# scale distance so that distances > tolerance become 255
alpha = 255 * distance / tolerance

# attach the alpha and save
newImage = image.bandjoin(alpha)

newImage.write_to_file("x.png")
```

and this is the image i am trying this on (baseImage.png):

![2021-10-11-10-17-7](https://user-images.githubusercontent.com/15367507/136756371-6c9c4673-cdd3-420d-adc6-cba3d27b7c5e.png)

**Comment by jcupitt** [comment, marked as answer]
Hi @ArtemAlagizov,

You're almost there -- you need to remove the old alpha. Try:

```python
newImage = image[0:3].bandjoin(alpha)
```

ie. take bands 0, 1, 2 (RGB) of the old image and attach the new alpha.

**Reply by alagiz** [reply]
great, thanks very much! works perfectly

**Reply by jcupitt** [reply]
You could make it quite a bit quicker by removing the square root, fwiw.

**Reply by sergeevabc** [reply]
@jcupitt Sorry, but how to change black to transparent with regular vips?

**Reply by jcupitt** [reply]
Hi @sergeevabc, do you mean the command-line interface? You just repeat the steps in the python code above ^^, but using temporary files instead of variables. But it'll be slow and annoying to code up -- the vips CLI is only useful for very basic tasks, I would use something like python / ruby / php / etc. for more complex jobs.

**Reply by sergeevabc** [reply]
@jcupitt I mean something like `magick.exe in.png -transparent #000000 out.png` but with `vips.exe`.

**Reply by jcupitt** [reply]
Something like that would be best in python. I wouldn't use the vips CLI.

Maybe:

```
vips relational_const blob.png t1.v equal "0 0 0 255"
vips bandbool t1.v t2.v and
vips extract_band blob.png t3.v 3
vips boolean t3.v t2.v t4.v eor
vips extract_band blob.png t5.v 0 --n 3
vips bandjoin "t5.v t4.v" x.png
```

But that's pretty terrible. I wouldn't do it.
